app/ota_client.py

```diff
@@ -896,48 +892,138 @@ def _gen_regular_files(self, rootfs_dir, regulars_file, target_dir):
-                                # /boot directory file
-                                logger.debug(f"boot file: {regular_inf.path}")
-                                self._gen_boot_dir_file(
-                                    rootfs_dir, target_dir, regular_inf, prev_inf, fl
-                                )
-                            else:
-                                # others
-                                logger.debug(f"no boot file: {regular_inf.path}")
-                                self._gen_regular_file(
-                                    rootfs_dir, target_dir, regular_inf, prev_inf, fl
-                                )
-                            prev_inf = regular_inf
-                    except Exception as e:
-                        logger.exception("regular files setup error!:")
-                        res = False
-                    finally:
-                        os.chdir(cwd)
-        shutil.copy(log_file, "./regulars_dl.log")
-        return res
+        
+        cwd = os.getcwd()
+        os.chdir(target_dir)
+
+        rfiles_list = []
+        with open(regulars_file) as f:
+            logger.debug(f"target_dir: {target_dir}")
+            rfiles_list = [RegularInf(l) for l in f.readlines()]
+
+        # process all regular files here
+        try:
+            self._process_regular_files(rootfs_dir, rfiles_list, target_dir)
+        except Exception as e:
+            logger.exception("regular files setup error!:")
+            return False
+        finally:
+            os.chdir(cwd)
+
+        return True
+
+    def _process_regular_files_pool_init(self, gvar_dict):
+        ''' 
+            Used by _process_regular_files
+            Init the worker pool with shared variables
+
+            DO NOT call this method from other functions 
+            except for _process_regular_files
+        '''
+        global global_var_dict
+        global_var_dict = gvar_dict
+
+    def _process_regular_files_exit(self, gvar_dict):
+        '''
+            Used by _process_regular_files
+            Update corresponding class attributes
+
+            DO NOT call this method from other functions 
+            except for _process_regular_files
+        '''
+        for k, v in gvar_dict:
+            # apply staged data to target attribute
+            vname = k.split("-")
+            if vname[0] == "staging":
+                if vname[1] == "dict":
+                    # TODO: update or replace?
+                    setattr(self, vname[-1], dict(v))
+                elif vname[1] == "list":
+                    setattr(self, vname[-1], list(v))
+
+    def _process_regular_files_log2file_listener(self, flog_path, log_queue, event):
+        '''
+            logger worker for _process_regular_files
+        '''
+        with open(flog_path, "w") as f:
+            while not event.is_set() or not log_queue.empty():
+                    msg = log_queue.get()
+                    f.write(msg+"\n")
```

Comment:
Can be removed this function and p_logger process.

Reply:
Resolve in commit 90f2727.